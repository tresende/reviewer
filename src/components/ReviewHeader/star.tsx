import { FaRegStar, FaStar } from 'react-icons/fa'
import theme from 'styles/theme'

export type StarProps = {
  outline?: boolean
}

const Star = ({ outline }: StarProps) => {
  const props = { color: theme.colors.yellow, size: 28 }
  return outline ? <FaRegStar {...props} /> : <FaStar {...props} />
}

export default Star
