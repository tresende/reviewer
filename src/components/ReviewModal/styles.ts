import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.div``

export const Header = styled.div`
  display: flex;
  flex-direction: column;
  height: 10rem;
  align-items: center;
`

export const HeaderBackground = styled.div`
  ${({ theme }) => css`
    width: 100%;
    position: relative;
    height: 7rem;
    border-radius: ${theme.border.mediumRadius} ${theme.border.mediumRadius} 0 0;
    padding: ${theme.spacings.small};
    background: linear-gradient(to top, rgba(250, 217, 97, 0.65), rgba(247, 107, 28, 0.65));
  `}
`

export const Avatar = styled.img`
  border-radius: 50%;
  border: 3px solid white;
  width: 8rem;
  height: 8rem;
  position: absolute;
  top: 2rem;
`

export const Body = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    min-height: 40rem;
    justify-content: space-between;
    padding: ${theme.spacings.medium};
    ${media.lessThan('small')`
      padding-bottom: 0;
    `}
  `}
`

export const Title = styled.span`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xlarge};
    font-weight: ${theme.font.light};
    span {
      color: ${theme.colors.gray450};
    }
  `}
`

export const SocialMedia = styled.div``

export const Textarea = styled.textarea`
  ${({ theme }) => css`
    width: 100%;
    height: 6em;
    padding: 6px 12px;
    border-radius: ${theme.border.smallRadius};
  `}
`
