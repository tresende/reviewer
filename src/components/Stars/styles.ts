import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    margin-right: ${theme.spacings.small};
  `}
`
