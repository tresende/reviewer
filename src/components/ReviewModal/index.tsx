import Body from './body'
import { save } from 'services/review'
import UserReview from 'models/UserReview'
import { useLoading } from 'hooks/use-loading'

import * as S from './styles'

const ReviewModal = () => {
  const loading = useLoading()

  const handleSubmit = async (data: UserReview) => {
    try {
      const text = 'Aguarde seu comentÃ¡rio estÃ¡ sendo enviado ğŸš€ğŸš€ğŸš€'
      loading.start(text)
      await save(data)
    } catch (error) {
      console.warn(error)
    } finally {
      loading.stop()
    }
  }

  return (
    <S.Wrapper>
      <S.Header>
        <S.HeaderBackground />
        <S.Avatar src="https://d3p07km83uit4h.cloudfront.net/portal/img/profile/avatar-blank.png?v=1051" />
      </S.Header>
      <Body handleSubmit={handleSubmit} />
    </S.Wrapper>
  )
}

export default ReviewModal
