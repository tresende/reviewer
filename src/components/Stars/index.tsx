import { useState } from 'react'

import Star from './star'
import { SCORE_LABEL, MAX_RATING } from 'utils/constants'

import * as S from './styles'

export type StarsProps = {
  quantity?: number
  size?: number
  onClick?: (index: number) => void
}

const Stars = ({ quantity = 1, size = 28, onClick }: StarsProps) => {
  const [selected, setSelected] = useState(quantity)

  const handleClick = (index: number) => {
    if (!onClick) return
    setSelected(index)
    onClick(index)
  }

  const solid = [...Array(selected)]
  const outline = [...Array(MAX_RATING - selected)]

  return (
    <S.Wrapper title={SCORE_LABEL[selected - 1]}>
      {solid.map((_, index) => (
        <Star key={index} onClick={() => handleClick(index + 1)} size={size} />
      ))}
      {outline.map((_, index) => {
        const currentIndex = index + 1 + selected
        return <Star key={currentIndex} onClick={() => handleClick(currentIndex)} outline size={size} />
      })}
    </S.Wrapper>
  )
}

export default Stars
