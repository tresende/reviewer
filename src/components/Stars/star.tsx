import { FaRegStar, FaStar } from 'react-icons/fa'

import theme from 'styles/theme'

import * as S from './styles'

export type StarProps = {
  onClick?: () => void
  outline?: boolean
  size: number
}

const Star = ({ onClick, outline, size }: StarProps) => {
  const props = { color: theme.colors.yellow, size }

  return <S.Button onClick={onClick}>{outline ? <FaRegStar {...props} /> : <FaStar {...props} />}</S.Button>
}

export default Star
