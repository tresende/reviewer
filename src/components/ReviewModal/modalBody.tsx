import { useState } from 'react'

import Stars from 'components/Stars'
import Button from 'components/Button'
import UserReview from 'models/UserReview'
import { MIN_TEXT_LENGTH, DEFAULT_SCORE } from 'utils/constants'

import * as S from './styles'

type BodyProps = {
  handleSubmit: (data: UserReview) => void
}
const INITIAL_STATE = { score: DEFAULT_SCORE, text: '' } as UserReview

const Body = ({ handleSubmit }: BodyProps) => {
  const [formData, setFormData] = useState<UserReview>(INITIAL_STATE)
  const [formEnable, setFormEnable] = useState(false)

  const handleChange = (field: string, value: string | number) => {
    const data = { ...formData, [field]: value }
    setFormData(data)
    setFormEnable(data?.text.length >= MIN_TEXT_LENGTH)
  }

  const onSave = () => {
    handleSubmit(formData)
    setFormData(INITIAL_STATE)
  }

  return (
    <S.Body>
      <S.Title>
        Ol치 <span>Thiago</span>, conte para comunidade como foi sua experi칡ncia em <span>BSB Coworking</span>.
      </S.Title>
      <Stars quantity={formData.score} size={52} onClick={(index) => handleChange('score', index)} />
      <span>Sua avalia칞칚o e coment치rio ajuda a comunidade ;)</span>
      <S.Textarea
        maxLength={280}
        onChange={(e) => handleChange('text', e.currentTarget.value)}
        placeholder="Escreva um coment치rio..."
      />
      <S.SocialMedia>
        <input id="twitter" type="checkbox" onChange={(e) => handleChange('twitter', e.currentTarget.value)} />
        <label htmlFor="twitter"> Desejo compartilhar no twitter 游닉</label>
      </S.SocialMedia>
      <div>
        <Button onClick={onSave} disabled={!formEnable} fillType="solid">
          Envie sua avalia칞칚o
        </Button>
      </div>
    </S.Body>
  )
}

export default Body
