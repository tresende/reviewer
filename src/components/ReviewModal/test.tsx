import userEvent from '@testing-library/user-event'
import { StarProps } from 'components/Stars/star'
import UserReview from 'models/UserReview'
import { render, screen } from 'utils/test-utils'
import ReviewModal from '.'

const mockService = jest.fn()
jest.mock('services/review', () => ({
  save: function Mock(data: UserReview) {
    if (data.text === 'FAIL TEST') throw Error('ERROR')
    mockService()
  }
}))

jest.mock('components/Stars/star', () => ({
  __esModule: true,
  default: function Mock(props: StarProps) {
    return <div onClick={props.onClick} data-testid={`star-${props.outline ? 'outline' : 'solid'}`} />
  }
}))

describe('<ReviewModal />', () => {
  it('should render component correctly', () => {
    const { container } = render(<ReviewModal />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('shouldnt call service when text > 5 caracteres', async () => {
    render(<ReviewModal />)

    await userEvent.type(screen.getByRole('textbox'), '1234')
    screen.getByRole('button', { name: /Envie sua avaliação/i }).click()

    expect(mockService).toBeCalledTimes(0)
  })

  it('shouldnt call service when text < 5 caracteres or more', async () => {
    render(<ReviewModal />)

    screen.getAllByTestId('star-solid')[0].click()
    await userEvent.type(screen.getByRole('textbox'), '12345')
    screen.getByRole('button', { name: /Envie sua avaliação/i }).click()
    expect(mockService).toBeCalledTimes(1)
  })

  it('shouldnt log error when api is down', async () => {
    console.warn = jest.fn()
    render(<ReviewModal />)

    await userEvent.type(screen.getByRole('textbox'), 'FAIL TEST')
    screen.getByRole('button', { name: /Envie sua avaliação/i }).click()
    expect(console.warn).toBeCalledTimes(1)
  })
})
