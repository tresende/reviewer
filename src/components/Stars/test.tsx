import { render, screen } from 'utils/test-utils'
import Stars from '.'
import Star, { StarProps } from './star'

jest.mock('components/Stars/star', () => ({
  __esModule: true,
  default: function Mock(props: StarProps) {
    return <div onClick={props.onClick} data-testid={`star-${props.outline ? 'outline' : 'solid'}`} />
  }
}))

describe('<Stars />', () => {
  it('should render component correctly', () => {
    const { container } = render(<Stars />)
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.getAllByTestId('star-solid')).toHaveLength(1)
  })

  it('should render 3 starts', () => {
    render(<Stars quantity={3} />)
    expect(screen.getAllByTestId('star-outline')).toHaveLength(2)
  })

  it('should execute callback on click in star', () => {
    const onClick = jest.fn()
    render(<Stars onClick={onClick} />)
    screen.getByTestId('star-solid').click()
    expect(onClick).toBeCalledTimes(1)
  })

  it('should ignore callback when onclick is null', () => {
    render(<Stars />)
    screen.getByTestId('star-solid').click()
  })

  describe('<Star />', () => {
    it('should render component correctly', () => {
      const { container } = render(<Star size={1} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
