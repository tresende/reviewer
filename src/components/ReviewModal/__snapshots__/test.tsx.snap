// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewModal /> should render component correctly 1`] = `
.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 2.4rem;
}

.c7 {
  padding: 0.6rem 1.2rem;
  width: 100%;
  border-radius: 0.4rem;
  border: 1px solid #FABA40;
  font-size: 1.8rem;
  box-shadow: 0px 8px 15px rgba(0,0,0,0.1);
  font-weight: 700;
  cursor: pointer;
  color: #FFFFFF;
  background: #FABA40;
}

.c7:disabled {
  cursor: not-allowed;
  -webkit-filter: saturate(30%);
  filter: saturate(30%);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 10rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 100%;
  position: relative;
  height: 7rem;
  border-radius: 0.8rem 0.8rem 0 0;
  padding: 2.4rem;
  background: linear-gradient(to top,rgba(250,217,97,0.65),rgba(247,107,28,0.65));
}

.c2 {
  border-radius: 50%;
  border: 3px solid white;
  width: 8rem;
  height: 8rem;
  position: absolute;
  top: 2rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  height: 36rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3.2rem;
}

.c4 {
  font-size: 2.0rem;
  font-weight: 300;
}

.c4 span {
  color: #7D7D7D;
}

.c6 {
  width: 100%;
  height: 6em;
  padding: 6px 12px;
  border-radius: 0.4rem;
}

@media (max-width:450px) {
  .c3 {
    padding-bottom: 0;
    height: 40rem;
  }
}

<div
  class=""
>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
    <img
      class="c2"
      src="https://d3p07km83uit4h.cloudfront.net/portal/img/profile/avatar-blank.png?v=1051"
    />
  </div>
  <div
    class="c3"
  >
    <span
      class="c4"
    >
      Olá 
      <span>
        Thiago
      </span>
      , conte para comunidade como foi sua experiência em 
      <span>
        BSB Coworking
      </span>
      .
    </span>
    <div
      class="c5"
      title="Mediano"
    >
      <div
        data-testid="star-solid"
      />
      <div
        data-testid="star-solid"
      />
      <div
        data-testid="star-solid"
      />
      <div
        data-testid="star-outline"
      />
      <div
        data-testid="star-outline"
      />
    </div>
    <span>
      Sua avaliação e comentário ajuda a comunidade ;)
    </span>
    <textarea
      class="c6"
      maxlength="280"
      placeholder="Escreva um comentário..."
    />
    <div
      class=""
    >
      <button
        class="c7"
        disabled=""
      >
        Envie sua avaliação
      </button>
    </div>
  </div>
</div>
`;
